---
title: "Introduction to MSstatsPTM"
author: "Tsung-Heng Tsai (tsai.tsungheng@gmail.com)"
output:
  BiocStyle::html_document:
    toc_float: TRUE
  pdf_document: default
package: MSstatsPTM
vignette: >
  %\VignetteIndexEntry{Introduction to MSstatsPTM}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r style, echo=FALSE, results='asis'}
BiocStyle::markdown()
```

```{r setup, echo=FALSE, message=FALSE}
library(MSstatsPTM)
```

# Introduction

`r Githubpkg("tsunghengtsai/MSstatsPTM")` provides a set of general statistical methods for characterization of quantitative changes in global post-translational modification (PTM) profiling experiments. Typically, the analysis involves quantification and integration of PTMs and their corresponding proteins.

Quantative analyses of PTMs are supported by four main function of _MSstatsPTM_:

* `PTMnormalize()` normalizes the quantified peak log2-intensities to correct systematic variation across MS runs.

* `PTMsummarize()` summarizes log2-intensities of spectral features into one value per PTM site (or protein) per MS run.

* `PTMestimate()` takes as input the summarized log2-intensities for each PTM site, performs statistical modeling for the abundance of the site, and returns the estimates of model parameters for all sites in all experimental conditions.

* `PTMcompareMeans()` performs significance analysis for detecting changes in PTM mean abundances between conditions.

# Installation

The development version of `r Githubpkg("tsunghengtsai/MSstatsPTM")` can be installed from GitHub:

```{r install, eval=FALSE}
devtools::install_github("tsunghengtsai/MSstatsPTM")
```

Once intalled, _MSstatsPTM_ can be loaded with `library()`:

```{r lib, eval=FALSE}
library(MSstatsPTM)
```

# Data preparation

The PTM abundance is affected by two factors: (1) the proportion of proteins carrying the PTM, and (2) the underlying protein abundance. Quantification of PTMs alone cannot provide full information about the degree of the modification. Therefore, a quantitative PTM experiment often involves analysis of both enriched samples (for PTMs) and unenriched samples (for proteins). 

The _MSstatsPTM_ analysis workflow takes as input a list of two data frames, named `PTM` and `Protein`. 

## Example dataset

We use `PTMsimulateExperiment()` to generate an example dataset.

```{r simulate}
sim <- PTMsimulateExperiment(
  nGroup = 2, nRep = 2, nProtein = 1, nSite = 2, nFeature = 5, 
  mu = list(PTM = 25, Protein = 25), 
  delta = list(PTM = c(0, 1), Protein = c(0, 1)), 
  sRep = list(PTM = 0.2, Protein = 0.2), 
  sPeak = list(PTM = 0.05, Protein = 0.05)
)
```

The function takes in account several parameters in a PTM experiment: number of groups (`nGroup`), number of replicates per group (`nRep`), number of proteins (`nProtein`), number of sites per protein (`nSite`), number of spectral features per site (`nFeature`), mean log2-abundance of PTM and protein (`mu`), deviation from the mean log2-abundance in each group (`delta`), standard deviation among replicates (`sRep`), and standard deviation among log2-intensities (`sPeak`), and creates a list of two data frames named `PTM` and `Protein`. 

Note here we use the generic term "feature" to refer to precursor ions in DDA, fragments in DIA, or transitions in SRM. 

```{r data-structure}
str(sim)
```

The `PTM` data frame contains 6 columns representing the quantified `log2inty` of each `feature`, `site` and `protein`, in the corresponding `group` and `run`: 

```{r data-ptm}
sim[["PTM"]]
```

The `Protein` data frame includes the same columns except `site`:

```{r data-protein}
sim[["Protein"]]
```

## Site-level representation

A main difference of the PTM analysis from general proteomics analyses is the focus on the PTM sites (i.e., modified residues), where the basic analysis unit is a PTM site, rather than a protein. A site can be spanned by multiple peptides, quantified with multiple spectral features. Transformation from peptide-level representation to site-level representation requires locating PTM sites on protein sequence, usually provided in a FASTA format. 

_MSstatsPTM_ provides several help functions to facilitate the transformation: 

* `tidy_fasta()` reads and returns the sequence information in a tidy format.

* `PTMlocate()` annotates PTM sites with the associated peptides and protein sequences. 

### Working with FASTA files

`tidy_fasta()` takes as input the path to a FASTA file (either a local path or an URL). For example, we can access to the information of Alpha-synuclein (P37840) on [uniprot](https://www.uniprot.org/uniprot/P37840) and extract the sequence information as follows:

```{r fasta}
tidy_fasta("https://www.uniprot.org/uniprot/P37840.fasta")
```


# Normalization

The normalization can be performed by equalizing sensible summary statistics (such as the median of the log2-intensities) across MS runs, or by using a reference dervied from either spiked-in internal standard or other orthogonal evidence. The PTM data and the protein data are normalized separately, using `PTMnormalize()`.

## Normalization based on the summary statistics of each run

```{r normalization, message=FALSE}
normalized <- PTMnormalize(sim)
```

## Normalization with reference


# Summarization

Use `PTMsummarize()` to summarize feature log2-intensities into one value for PTM site per run. 

## Summarization options

There are four summarization options avaialble in _MSstatsPTM_: 

- `tmp` (default): Tukey's median polish procedure
- `logsum`: log2 of the summation of peak intensities
- `mean`: mean of the log2-intensities
- `median`: median of the log2-intensities
- `max`: max of the log2-intensities

# Estimation

Use `PTMestimate()` to perform the statistical model and inference of the abundance of each PTM site in each group.

# Detection of changes in PTMs

Use `PTMcompareMeans()` to detect changes in mean abundances of PTM sites

## Detection of changes in mean abundances of PTM sites



## Protein-level adjustment

Adjustment with respect to protein abundance
